pkg/root/trusted_root_create.go

```diff
@@ -0,0 +1,414 @@
+// Copyright 2023 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package root
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/hex"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	protocommon "github.com/sigstore/protobuf-specs/gen/pb-go/common/v1"
+	prototrustroot "github.com/sigstore/protobuf-specs/gen/pb-go/trustroot/v1"
+	timestamppb "google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	FulcioTarget = "Fulcio"
+	RekorTarget  = "Rekor"
+	CTFETarget   = "CTFE"
+	TSATarget    = "TSA"
+)
+
+type RawTrustedRootTarget interface {
+	GetType() string
+	GetBytes() []byte
+}
+
+type BaseTrustedRootTarget struct {
+	Type  string
+	Bytes []byte
+}
+
+func (b *BaseTrustedRootTarget) GetType() string {
+	return b.Type
+}
+
+func (b *BaseTrustedRootTarget) GetBytes() []byte {
+	return b.Bytes
+}
+
+// NewTrustedRootFromTargets initializes a TrustedRoot object from a mediaType string and
+// a slice of targets. These targets are expected to be PEM-encoded public keys/certificate chains.
+// This method of constructing the TrustedRoot has some shortcomings which can be aided by manually
+// adjusting the TrustedRoot object after instantiation:
+//
+//   - publicKey instances for tlogs/ctlogs will have validFor.start set to Time.now() and no validFor.end.
+//   - Merkle Tree hash function is hardcoded to SHA256, as this is not derivable from the public key.
+//   - Each certificate chain is expected to be given as a single target, where there is a newline
+//     between individual certificates. It is expected that the certificate chain is ordered (root last).
+func NewTrustedRootFromTargets(mediaType string, targets []RawTrustedRootTarget) (*TrustedRoot, error) {
```

Comment:
I think there's a lot of use-cases for making it easier to construct a trusted root, and the challenge here is going to be to make an API that's flexible enough to cover those use-cases. If you're in something like `cosign verify-blob` with `--certificate-chain` and `--timestamp-certificate-chain`, it's totally fine to throw together something quickly to verify. But if you're the public good instance (or a private instance) you're going to want something that supports rotating key material with expiry times.

In `pkg/root/trusted_root.go` we already have these nice intermediate representations of `CertificateAuthority` and `TransparencyLog`, so what if we added there something like:

```
func NewTrustedRoot(
    certificateAuthorities []CertificateAuthorities, 
    certificateTransparencyLogs []TransparencyLog,
    timestampAuthorities []CertificateAuthorities,
    transparencyLogs []TransparencyLog
) TrustedRoot
```

This API would be similar to `sign.BundleOptions`, which is nice.

So if you're in `cosign verify-blob` and all you have is a PEM-encoded byte string from `--certificate-authority`, you could construct a minimal `CertificateAuthority` to pass in to `NewTrustedRoot`, but if you had a start and end time that certificate was valid for, you could also supply that information. But (crucially!) then we don't have to guess as to what the configuration should be, and we don't risk returning a half-configured object that the caller then has to go and fix up.

That would be enough if you just need a `TrustedRoot` to call into `sigstore-go`, but if you want to output a JSON file we might need one more step. I love the idea of adding a function like:

```
func (tr *TrustedRoot) MarshalJSON() ([]byte, error)
```

but instead of the current implementation of `constructProtoTrustRoot()`, maybe we could see if `TrustedRoot.trustedRoot` is already set up (like if it came from `NewTrustedRootFromProtobuf()`) or not (like if it came from the proposed `NewTrustedRoot()`) and only populate content in `TrustedRoot.trustedRoot` if it hasn't already been set up. This would ensure that if you load from a JSON file and call `MarshalJSON()` you get the same content back.

Does that make sense? It might make sense to get feedback on the API and approach from a few folks and settle on a direction before we move ahead with the implementation.

Reply:
Thanks for your input! I've been thinking about this and I have this alternative implementation proposal:

* Instead of having the current `NewTrustedRootFromTargets` function, I will add `NewTrustedRoot` as you suggested.
* I will also expose functions that can instantiate `CertificateAuthority` and `TransparencyLog` from certificate/public key material. These functions already exist in this PR, but are private. I'll revisit them and ensure their API is sane, name them better etc, so that they can be exposed. This will make sure that my usecase of easily instantiating the whole `TrustedRoot` object from the certificate/public key material is still very easy using this library.
* In terms of the `MarshalJSON` call, it's important to note that we publicly allow users to access *and modify* the structures that form the `TrustedRoot` object via `TimestampingAuthorities`, `FulcioCertificateAuthorities` functions etc. This means that if we just use the current `trustedRoot`, we could be serializing an outdated representation of `TrustedRoot`.

Does that make sense? Also CC @haydentherapper, I think this should also address your original concern. Let me know if this makes sense.

Reply:
> I will also expose functions that can instantiate CertificateAuthority and TransparencyLog from certificate/public key material.

I think any helper functions like this should live outside of sigstore-go. It might seem silly, but one of the design goals of sigstore-go is to have opinionated interfaces that get wrapped by things that call sigstore-go to be more user-friendly. A great example of this is cosign, where you can supply information via a URL, a path on disk, a string of encoded data, etc. We want that flexibility to live in cosign, instead of having several similar functions in sigstore-go that support slightly different data input.

> In terms of the MarshalJSON call, it's important to note that we publicly allow users to access and modify the structures that form the TrustedRoot object via TimestampingAuthorities, FulcioCertificateAuthorities functions etc. This means that if we just use the current trustedRoot, we could be serializing an outdated representation of TrustedRoot.

Got it - makes sense! Like you say, we should make sure whatever we marshal reflects the current state of the object.

Reply:
> I think any helper functions like this should live outside of sigstore-go. It might seem silly, but one of the design goals of sigstore-go is to have opinionated interfaces that get wrapped by things that call sigstore-go to be more user-friendly. A great example of this is cosign, where you can supply information via a URL, a path on disk, a string of encoded data, etc. We want that flexibility to live in cosign, instead of having several similar functions in sigstore-go that support slightly different data input.

I certainly don't want to go against the design goal that you folks have. TBH I wasn't even aware that this was the design goal - is there a good document that outlines design goals?

The only reason why I opened this PR in here was that @haydentherapper suggested it in https://github.com/sigstore/scaffolding/pull/1191#discussion_r1692119811. I can definitely change the PR in the way you suggested, but I would like Hayden to comment on here to ensure I'm doing changes that everybody will agree with conceptually.